### PR TITLE
Add Pagination for Posts

### LINKED ISSUE #1046: Pagination in Posts Block
### Description:
Exploring if having pagination for the Posts Block will be a nice addition. I can imagine two ways of doing it. PHP based where the page will reload and show the pagination based on the query string or we can load a JS file and do it client-side.

Feedback is appreciated if you think the post block that we have should have a pagination option or if you think that's quite outside the scope for a block that it more like a recent posts block.

@Codeinwp/design-team 
### Alternatives:

 
Not to have it and closing this ticket happily ever after.

### PR DESCRIPTION
**Pull request by Soare-Robert-Daniel**
Closes #1046

### Summary

Add pagination for Posts Block.

Developing notes:
 - It seems that Dynamic Content collides with the counting mechanism. It needs some investigation on the cause.

### Screenshots 

image

image

https://github.com/Codeinwp/otter-blocks/assets/17597852/9bdbff2e-11cb-457d-8271-2a2572167e12

----

### Test instructions

1. Insert a Post Block
2. Enable pagination and customize it.

---- 

### Checklist before the final review

- [ ] Included E2E or unit tests for the changes in this PR.
- [x] Visual elements are not affected by independent changes.
- [x] It is at least compatible with the [minimum WordPress version](https://wordpress.org/plugins/otter-blocks/).
- [x] It loads additional script in frontend only if it is required.
- [x] Does not impact the [Core Web Vitals](https://web.dev/vitals/).
- [x] In case of deprecation, old blocks are safely migrated.
- [x] It is usable in Widgets and FSE.
- [x] Copy/Paste is working if the attributes are modified.
- [x] PR is following [the best practices]()

### CODE REVIEW

File: inc/render/class-posts-grid-block.php [hunk 1]
```diff
@@ -351,25 +345,99 @@ protected function render_featured_post( $post, $attributes ) {
...
+			$posts = array_merge( $posts, get_posts( $args ) );
+
+			if ( $count_posts ) {
+				unset( $args['offset'] );
+				unset( $args['numberposts'] );
+				$args['posts_per_page'] = -1;
+				$args['fields']         = 'ids';
+
+				$total_posts += ( new \WP_Query( $args ) )->found_posts;
+			}
```
Comment (Soare-Robert-Daniel) [inline inc/render/class-posts-grid-block.php:412]: @HardeepAsrani, any idea if we can do the page counting to be nicer?

**Review by HardeepAsrani — CHANGES_REQUESTED** [review]
Hey, I was just testing it on an instance to see the flow and pagination doesn't work:

Screenshot 2023-10-04 at 10 17 39 AM
Screenshot 2023-10-04 at 10 17 32 AM

Regardless the page, it loads the same posts.

**Comment by Soare-Robert-Daniel** [comment]
@HardeepAsrani, did you test on WPTaste? For some reasons it tries to always redirect me to the first page.

File: inc/render/class-posts-grid-block.php [hunk 2]
```diff
@@ -350,26 +314,109 @@ protected function render_featured_post( $post, $attributes ) {
...
-	 * @return array|array[]|int[]|null[]|\WP_Post[] Posts.
+	 * @param array $attributes Blocks attrs.
+	 * @param bool  $count_posts Enable post count.
+	 * @param int   $page_number Page number.
+	 * @param int   $total_posts Total posts.
+	 * @return array|int[]|null[]|\WP_Post[] Posts.
 	 */
-	protected function get_posts( $args ) {
-		if ( isset( $args['post_type'] ) && 'product' === $args['post_type'] && function_exists( 'wc_get_products' ) ) {
+	protected function retrive_posts( $attributes, $count_posts, $page_number, &$total_posts ) {
```
Comment (HardeepAsrani) [inline inc/render/class-posts-grid-block.php:323]: NIT: You can fix the type, it should be `retrieve_posts`

**Review by HardeepAsrani — APPROVED** [review]
Left a small NIT but it looks good. It worked well when tried it locally. We can put it through QA and remove the QA-related code afterwards.

File: inc/render/class-posts-grid-block.php [hunk 3]
```diff
@@ -159,14 +117,20 @@ function ( $x ) use ( $sticky_posts_id ) {
...
 			trim( $class ),
-			$list_items_markup
+			$list_items_markup,
+			$has_pagination ? $this->render_pagination( $page_number, $total_posts ) : ''
 		);
 
+		// TODO: Remove after QA.
+		$end_time       = microtime( true );
+		$elapsed_time   = $end_time - $start_time;
+		$block_content .= '<!-- Rendered in ' . $elapsed_time . ' seconds | Page ' . $page_number . ' -->';
```
[the remark this answers is not in the document]
Comment (HardeepAsrani) [reply inc/render/class-posts-grid-block.php:132]: Might be worth testing this on our end with a dummy website with 1000s of posts, but we don't need to put it in prod given the number doesn't make much sense unless you can test before/after results.

**Comment by irinelenache** [comment]
@Soare-Robert-Daniel Tested the feature and here is what i found:

- [ ] **Pagination active border** color control is duplicated https://vertis.d.pr/i/55HbEu
- [ ] There is no way to add a top margin for the pagination, so if the row gap is set to a higher number, it won't look good https://vertis.d.pr/i/Tm19RT
- [ ] The featured post will be different on every page, both times when it's set as **Sticky post** or **Latest post** https://vertis.d.pr/v/f5XGgA
- [ ] If i enable the featured post and the number of items is set to 4, there will be 5 posts on the first page and 4 posts on the rest of them: [first page](https://vertis.d.pr/i/uII70G), [second page](https://vertis.d.pr/i/Sb5ioq)
- [ ] If the posts block is not placed at the top of the page, switching through pagination will refresh the page and it won't be scrolled to the block. Maybe we can change this for a better UX
- [ ] If i use [these permalink settings](https://vertis.d.pr/i/Wa9MXX), the pagination won't work https://vertis.d.pr/i/to87al (The block is placed on the homepage)

You can check the issues on this instance:
```
 Admin area URL: https://cycleadvertisement.s2-tastewp.com/wp-admin 
 Username: irinel 
 Password: hQ688Q6QlTQ 
```

**Comment by Soare-Robert-Daniel** [comment]
> * [ ]  The featured post will be different on every page, both times when it's set as **Sticky post** or **Latest post** https://vertis.d.pr/v/f5XGgA
> * [ ]  If i enable the featured post and the number of items is set to 4, there will be 5 posts on the first page and 4 posts on the rest of them: [first page](https://vertis.d.pr/i/uII70G), [second page](https://vertis.d.pr/i/Sb5ioq)

For those problems, I made the pagination and features most to be mutually exclusive. You will not have them both. The way the featured post works right now is like a stylized first post. This is not good in the context of pagination in which the featured post must be a separate thing.

If we want them both, we will make another issue to address this. In this PR we will focus only on pagination. 

> * [ ]  If the posts block is not placed at the top of the page, switching through pagination will refresh the page and it won't be scrolled to the block. Maybe we can change this for a better UX

This is a hard problem because when you enter into the page, we do not know if you come via pagination or other ways around.  It creates some weird behavior. We will mark it as not an issue since some other competition blocks have the same behavior.

**Comment by Soare-Robert-Daniel** [comment]
@irinelenache, when testing, please use another instance and not the one in this PR.

**Comment by irinelenache** [comment]
@Soare-Robert-Daniel Thank you for the fixes and for clarifying 🙏 

> If i use [these permalink settings](https://vertis.d.pr/i/Wa9MXX), the pagination won't work https://vertis.d.pr/i/to87al (The block is placed on the homepage)

^ This issue is still happening on my end, you can check on this instance:
```
Site name: Tranquil Division
URL: https://tranquildivision.s2-tastewp.com
Username: irinel
Password: fUuJjouVI7I
```

Besides this, everything's working fine 👍

File: inc/render/class-posts-grid-block.php [hunk 4]
```diff
@@ -350,26 +307,111 @@ protected function render_featured_post( $post, $attributes ) {
...
+			array(
+				'base'      => $base,
+				'format'    => $format,
+				'current'   => $page_number,
+				'total'     => $total_pages,
+				'prev_text' => __( 'Prev', 'otter-blocks' ),
+				'next_text' => __( 'Next', 'otter-blocks' ),
+			)
+		);
+		$output .= '</div>';
```
Comment (Soare-Robert-Daniel) [inline inc/render/class-posts-grid-block.php:413]: @HardeepAsrani, do you see something wrong with this? When added to Home page, the link is something like this `https://cycleadvertisement.s2-tastewp.com/page/2/` and is not working. If you put the block on a post page, the link will be like `https://cycleadvertisement.s2-tastewp.com/index.php/2023/10/12/post/page/2/` and is working.

Comment (HardeepAsrani) [reply inc/render/class-posts-grid-block.php:413]: @Soare-Robert-Daniel When you're going to that URL, I suppose it's redirecting you to 404? Or is it still opening the home?

Also, is it the sam eissue with Qubely?

Comment (HardeepAsrani) [reply inc/render/class-posts-grid-block.php]: Might be some issue with nginx/htaccess configs maybe, then we can skip that part and live with it for this version.

**Comment by Soare-Robert-Daniel** [comment]
@irinelenache, we should skip the front page on WP Taste since it might differ on host settings. We should use this plugin https://wordpress.org/plugins/qubely/ as a reference. Like Otter, it uses the same functions, so where they work, we should also work. The vice versa is also true.

**Comment by irinelenache** [comment]
@Soare-Robert-Daniel 
I tried to test on InstaWP now and compared some behaviours with Qubely. If i'm using these [permalink settings](https://vertis.d.pr/i/0NwtQE) it will work fine on the homepage only. On a normal page, there will be a slash missing that will break the page, here is a video with the behaviour on our end compared to Qubely https://vertis.d.pr/v/jyUFuS

**Comment by Soare-Robert-Daniel** [comment]
I made the code similar to on Quble, which uses a snippet from PHP docs -- there should be no difference now. Here are some explanations for the trick used: https://chat.openai.com/share/3af75654-25bb-40de-988f-21ac8413b9e0

**Comment by irinelenache** [comment]
@Soare-Robert-Daniel Checked again and everything's fine now, thank you 🚀

**Pull request merged**